File: src/view/ScreenImplConsole.java
package view;

import model.structs.Point;
import model.structs.Rect;

import java.util.ArrayList;

public class ScreenImplConsole implements Screen {
    private final Pixel[][] buffer;
    private final Rect rect;
    private final ArrayList<Window> windows;
    private int dirty;

    public ScreenImplConsole(int height, int width) {
        this.dirty = 0;
        this.rect = new Rect(0, height, 0, width);
        this.buffer = new Pixel[rect.bottom][rect.right];
        this.windows = new ArrayList<>();

        for(int i = rect.top; i < rect.bottom; i++) {
            for(int j = rect.left; j < rect.right; j++) {
                buffer[i][j] = new PixelImplConsole();
            }
        }
    }


    /**
     * get a view to draw on
     * @param rect the rect of view.
     * @return Window if allocated
     */
    @Override
    public Window createWindow(Rect rect) {
        Window window = new WindowImplConsole(rect, this);
        windows.add(window);
        return window;
    }

    @Override
    public Rect getRect() {
        return rect;
    }

    /**
     * set pixel on screen
     * set dirty to true if really changed
     *
     * @param position pixel position
     * @param pix the pixel 'color'
     * @return dirty since we don't care one pixel but the whole screen
     */
    protected int setPixel(Point position, Pixel pix) {
        dirty += buffer[position.y][position.x].set(pix);
        return dirty;
    }

    /**
     * display the screen, force refresh by default
     * @return true when repainted
     */
    @Override
    public boolean paint() {
        if(dirty == 0) {
            return false;
        }
        return forcePaint();
    }

    @Override
    public boolean forcePaint() {
        for(int i = rect.top; i < rect.bottom; i++) {
            for(int j = rect.left; j < rect.right; j++) {
                buffer[i][j].paint();
                buffer[i][j].flush();
            }
            System.out.print('\n');
        }
        dirty = 0;
        return true;
    }

    @Override
    public int clearScreenBuffer() {
        PixelImplConsole pixelBlank = new PixelImplConsole();
        for(int i = rect.top; i < rect.bottom; i++) {
            for(int j = rect.left; j < rect.right; j++) {
                buffer[i][j].set(pixelBlank);
            }
            System.out.print('\n');
        }
        return 0;
    }
}

File: src/view/console/ScreenImplConsole.kt
package view.console

import model.structs.Point
import model.structs.Rect
import view.console.Screen.Companion.clear

class ScreenImplConsole(height: Int, width: Int) : Screen {
    private val buffer: Array<Array<Pixel>> = Array(height) { Array(width) { PixelImplConsole() } }
    override val rect: Rect = Rect(0, height, 0, width)
    private val windows: ArrayList<Window> = ArrayList<Window>()
    private var dirty = 0

    /**
     * get a view to draw on
     * @param rect the rect of view.
     * @return Window if allocated
     */
    override fun createWindow(rect: Rect): Window {
        val window: Window = WindowImplConsole(rect, this)
        windows.add(window)
        return window
    }

    /**
     * set pixel on screen
     * set dirty to true if really changed
     *
     * @param position pixel position
     * @param pix the pixel 'color'
     * @return dirty since we don't care one pixel but the whole screen
     */
    fun setPixel(position: Point, pix: Pixel): Int {
        dirty += buffer[position.y][position.x].set(pix)
        return dirty
    }

    /**
     * display the screen, force refresh by default
     * @return true when repainted
     */
    override fun paint(): Boolean {
        if (dirty == 0) {
            return false
        }
        return forcePaint()
    }

    override fun forcePaint(): Boolean {
        clear()
        val sb = StringBuilder()
        sb.ensureCapacity(rect.right - rect.left)
        for (i in rect.top..<rect.bottom) {
            sb.setLength(0)
            for (j in rect.left..<rect.right) {
                sb.append(buffer[i][j].paint())
                buffer[i][j].flush()
            }
            println(sb)
        }
        dirty = 0
        return true
    }

    override fun clearScreenBuffer(): Int {
        val pixelBlank = PixelImplConsole()
        for (i in rect.top..<rect.bottom) {
            for (j in rect.left..<rect.right) {
                buffer[i][j].set(pixelBlank)
            }
        }
        return 0
    }
}
